pep-0684.rst

```diff
@@ -0,0 +1,640 @@
+to each interpreter (or a global lock will be needed).  This is the
+highest risk part of the work to isolate interpreters and may require
+more than just moving fields down from ``_PyRuntimeState``.  Some of
+the complexity is reduced if CPython switches to a thread-safe
+allocator like mimalloc.
+
+.. _proposed capi:
+
+C-API
+-----
+
+The following private API will be made public:
+
+* ``_PyInterpreterConfig``
+* ``_Py_NewInterpreter()`` (as ``Py_NewInterpreterEx()``)
+
+The following fields will be added to ``PyInterpreterConfig``:
+
+* ``own_gil`` - (bool) create a new interpreter lock
+  (instead of sharing with the main interpreter)
+* ``strict_extensions`` - fail import in this interpreter for
+  incompatible extensions (see `Restricting Extension Modules`_)
+
+Restricting Extension Modules
+-----------------------------
+
+Extension modules have many of the same problems as the runtime when
+state is stored in global variables.  :pep:`630` covers all the details
+of what extensions must do to support isolation, and thus safely run in
+multiple interpreters at once.  This includes dealing with their globals.
+
+Extension modules that do not implement isolation will only run in
+the main interpreter.  In all other interpreters the import will
+raise ``ImportError``.  This will be done through
+``importlib._bootstrap_external.ExtensionFileLoader``.
+
+We will work with popular extensions to help them support use in
+multiple interpreters.  This may involve adding new public APIs, which
+we will address on a case-by-case basis.
+
+Extension Module Compatibility
+''''''''''''''''''''''''''''''
+
+As noted in `Extension Modules`_, many extensions work fine in multiple
+interpreters without needing any changes.  The import system will still
+fail if such a module doesn't explicitly indicate support.  At first,
+not as many extension modules will do so, so this is a potential source
+of frustration.
+
+We will address this by adding a context manager to temporarily disable
+the check on multiple interpreter support:
+``importlib.util.allow_all_extensions()``.
+
+Documentation
+-------------
+
+The "Sub-interpreter support" section of Doc/c-api/init.rst will be
+updated with the added API.
+
+
+How to Teach This
+=================
+
+This is an advanced feature for users of the C-API.  There is no
+expectation that this will be taught.
+
+That said, if it were taught then it would boil down to the following::
+
+    In addition to Py_NewInterpreter(), you can use Py_NewInterpreterEx()
+    to create an interpreter.  The config you pass it indicates how you
+    want that interpreter to behave.
+
+
+Extra Context
+=============
+
+Sharing Global Objects
+----------------------
+
+We are sharing some global objects between interpreters.
+This is an implementation detail and relates more to
+`globals consolidation <Consolidating Runtime Global State>`_
+than to this proposal, but it is a significant enough detail
```

Comment:
```suggestion
than to this proposal, but it is significant enough
```

Avoid repetition

Reply:
The repetition was on purpose, for emphasis.